### PR TITLE
feat(Timestamp): create component

### LINKED ISSUE #7729: Timestamp - a new Timestamp component
Timestamp is a new component that will take a raw timestamp value and convert it into the desired display format.

Design: https://marvelapp.com/prototype/2cf6171g
Core issue: https://github.com/patternfly/patternfly/issues/4801

### PR DESCRIPTION
**Pull request by thatblindgeye**
**What**: Closes #7729

[Timestamp preview build](https://patternfly-react-pr-7815.surge.sh/components/timestamp)

Still to-do:
- [x] Add classes (currently commented out the line that imports Timestamp styles)
- [x] Add tests
- [ ] Possibly refactor some examples to utilize checkbox/select modifiers (i.e. the "Basic formats" example could use a single timestamp, and just have checkboxes/select menus for users to tinker with different prop values)

**Additional issues**:

### CODE REVIEW

**Review by mcoker — CHANGES_REQUESTED** [review]
Looks awesome to me! 2 things:

* Looks like we're missing the `__text` class on the `time` element. I'm assuming because that isn't showing up in react-styles since the class doesn't have any styles from core? In those instances, I believe what we do is hardcode the class so that it's there and we can reference it if we need without introducing a new class, but maybe one of the react devs can verify if that's still the case.
* There is some extra whitespace being created within the `<time>` element - at both the start and end of the text with custom content, and at the end of the regular examples. Ideally we would take that out, but particularly the text at the end because it's causing the timestamp (most evident in the tooltip examples, since there is a border) to contain the whitespace between it and whatever comes after it since only the first whitespace character is preserved and any add'l whitespace chars are discarded.

Here's what that looks like with inline content after the timestamp - I added a light background to make it more obvious, but you can also see the help text border extends past the text since it's applying to that whitespace character.

Screen Shot 2022-08-17 at 1 17 13 PM

Here's what it looks like if we take that extra whitespace out

Screen Shot 2022-08-17 at 1 18 59 PM

FWIW here's what it looks like when it's focused currently

Screen Shot 2022-08-17 at 1 20 23 PM

Here's what it looks like focused with the whitespace removed

Screen Shot 2022-08-17 at 1 20 53 PM

And since this is just inline text, the outline is a little tight. It's fine, but we could add an `outline-offset` to create some breathing room around the text. wdyt about the outline @mcarrano?

Screen Shot 2022-08-17 at 1 21 04 PM

File: packages/react-core/src/components/Timestamp/Timestamp.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+export enum TimestampFormats {
+  full = 'full',
+  long = 'long',
+  medium = 'medium',
+  short = 'short'
+}
+
+export interface TimestampProps extends React.HTMLProps<HTMLSpanElement> {
+  /** Custom content rendered inside the timestamp, such as a relative time. Passing this prop
+   * in will override the default content and timeZoneSuffix prop.
```
Comment (tlabaj) [inline packages/react-core/src/components/Timestamp/Timestamp.tsx:15]: There is no content prop so I would reward this so it is more clear that it overrides the default display string.

File: packages/react-core/src/components/Timestamp/Timestamp.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+   *
+   * For more information about the time element's datetime attribute:
+   * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/time#valid_datetime_values
+   */
+  datetime: string;
+  /** Flag indicating whether the timestamp has a tooltip to display the UTC time.
+   * To render a tooltip with custom content, pass in tooltipProps instead.
+   */
+  hasUTCTooltip?: boolean;
+  /** Flag for displaying the time in a 12 hour format (true) or a 24 hour format (false),
```
Comment (tlabaj) [inline packages/react-core/src/components/Timestamp/Timestamp.tsx:45]: The default values are show in the prop table.  I also don't think it is necessary to mention it is only used for custom content.  I think that is implied.   Maybe we can shorten this to:
"Flag for displaying the time in a 12 hour format. If this prop is not passed in, the hour format will be based on the locale prop's value."
Ee could also add a blurb somewhere explaining in implications of using children prop.

File: packages/react-core/src/components/Timestamp/Timestamp.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      className={css(styles.timestamp, hasTooltip && styles.modifiers.helpText, className)}
+      {...(hasTooltip && { tabIndex: 0 })}
+      {...props}
+    >
+      <time dateTime={datetime}>{!children ? `${dateAsLocaleString} ${timeZoneSuffix}` : children}</time>
+    </span>
+  );
+
+  return hasTooltip ? (
+    <Tooltip content={tooltipProps?.content || `${utcDateString} ${utcSuffix}`} {...tooltipProps}>
```
Comment (tlabaj) [inline packages/react-core/src/components/Timestamp/Timestamp.tsx:120]: I do not see an example of the tooltip in core.  I do however see a mockup in the design. The design mentions a Popover though.
I think we should use a similar API here as we do for the Progress stepper help text where we use the render prop and pass in the popover.

**Review by tlabaj — CHANGES_REQUESTED** [review]
Looking really good so far.

File: packages/react-core/src/components/Timestamp/Timestamp.tsx (hunk 3 above)
Comment (tlabaj) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:120]: I just saw conversation in issue.  SO it should be a tooltip.  I would still suggest a similar API to the progress stepper's help text.

File: packages/react-core/src/components/Timestamp/Timestamp.tsx (hunk 2 above)
Comment (thatblindgeye) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:45]: For the "when custom content is not passed in" text, that was more to convey that this prop will only really work when displaying the default content. I could reword it a bit to either say "...when not using the children prop", or "...when the default content is displayed." (along with other areas where I mention "when custom content is not passed in"). What do you think?

File: packages/react-core/src/components/Timestamp/examples/Timestamp.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+By default, a timestamp will display the current date and time based on the current locale if the `date` prop is not passed in.
+
+The following example shows one way to build an ISO 8601 string with the current local date and time, which gets passed into the `datetime` prop. Alternatively, the current UTC date and time could be passed in using the `toISOString` method.
+
+```ts file="./TimestampDefault.tsx"
+```
+
+### Basic formats
+
+The format of the displayed content can be customized by passing in the `dateFormat` and/or `timeFormat` props. Passing in only one of the props will display only the date or time, depending on which prop is passed in. The possible options are "full", "long", "medium", and "short".
```
[the remark this answers is not in the document]
Comment (thatblindgeye) [reply packages/react-core/src/components/Timestamp/examples/Timestamp.md:22]: > Looks like we're missing the __text class on the time element. I'm assuming because that isn't showing up in react-styles since the class doesn't have any styles from core? In those instances, I believe what we do is hardcode the class so that it's there and we can reference it if we need without introducing a new class, but maybe one of the react devs can verify if that's still the case.

That was my assumption as well when I saw that class not able to be applied via `styles...`. For now I hardcoded the class in.

> There is some extra whitespace being created within the <time> element - at both the start and end of the text with custom content, and at the end of the regular examples...

Good catch! I've updated locally how the default content gets rendered so the trailing whitespace at the end should not longer be there. In terms of what it looks like in dev tools, I noticed that the Title component examples look similar (with a white space between the `>` and `<` brackets). It doesn't look like that is affecting what is actually rendered at least now.

With the timeZoneSuffix:

timestamp with timezone suffix

Without the suffix:

timestamp without timezone suffix

For the outline, I agree it does look a bit tight. This is what it looks like with 0.25rem (I believe the global xs spacer CSS variable):

timestamp outline

I also had tried just using 2px and it looked okay as well, but the dashed underline gets cut off with that. If we want to add an outline offset I can open up a PR in Core to add that in.

File: packages/react-core/src/components/Timestamp/Timestamp.tsx (hunk 1 above)
Comment (tlabaj) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:15]: I think that is better.  I do not think it is necessary to mention the timeZoneSuffix prop.

File: packages/react-core/src/components/Timestamp/Timestamp.tsx (hunk 3 above)
Comment (thatblindgeye) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:120]: Originally that was what I was considering, but I ended up going with this approach to avoid using refs since there was previously an issue with tooltips via refs getting announced correctly (though tooltips with refs do work better now). I can update to be more similar to how the Progress Stepper uses Popover's. 

Based on conversation in #7729 , I had opted to have one prop for rendering a default, UTC tooltip, and another for rendering a custom tooltip. Do you feel having both would still be good (replacing the current `tooltipProps` prop with a `tooltipRender` prop that matches ProgressStepper)?

File: packages/react-core/src/components/Timestamp/Timestamp.tsx (hunk 2 above)
Comment (tlabaj) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:45]: I feel like most of the props do not apply to custom content so it seems like instead of mentioning it for every prop, we could write a blurb somewhere,  maybe even in the description of the children prop.

File: packages/react-core/src/components/Timestamp/Timestamp.tsx (hunk 3 above)
Comment (tlabaj) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:120]: I think it is best to be consistent if we can.  If the we truly can't make the tooltip accessible with the API we use elsewhere for help text, then we can consider something else.
That being said, since we want a  default tooltip here.  I still think we can improve the API a bit.  I think using `tooltipProps` for custom tooltips is little confusing.  Maybe we can combine the props 

```
export interface TimestampTooltipObject {
  /** variant of tooltip*/
  variant: 'default' | 'custom' ;
  /** custom content */
  content?: React.ReactNode;
 /** Additional tooltip props **/
  tooltipProps?: TooltipProps;
}
```
Or something like.  Then if the object exists you can render a tooltip.

**Review by tlabaj — CHANGES_REQUESTED** [review]

File: packages/react-core/src/components/Timestamp/examples/Timestamp.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (mcoker) [reply packages/react-core/src/components/Timestamp/examples/Timestamp.md:22]: @thatblindgeye go ahead and open a PR, that would be great - thanks! What we typically do in core is use an SCSS function to render the correct `rem` value.

> I noticed that the Title component examples look similar (with a white space between the > and < brackets). It doesn't look like that is affecting what is actually rendered at least now.

Yeah, I imagine that's a common thing. It isn't really visible and _probably_ not a thing unless there is some kind of indication of where the text/thing (like an image or icon) in the thing stops and starts - the reason I noticed it is because the text has the underline. The title component is a block element so it wouldn't have that issue, since it will push everything before/after it onto different lines. Also there isn't an indication of where it stops/starts unless we added a border or background or something.

A really common problem I notice all over the place is with links. It's visible in our text component, too:

Screen Shot 2022-08-18 at 10 38 22 AM


Screen Shot 2022-08-18 at 10 38 34 AM

File: packages/react-core/src/components/Timestamp/Timestamp.tsx [hunk 5]
```diff
@@ -77,47 +87,51 @@ export const Timestamp: React.FunctionComponent<TimestampProps> = ({
...
     ...(customFormat && { ...customFormat }),
     ...(is12Hour !== undefined && { hour12: is12Hour })
   };
-  const hasTimeFormat = timeFormat && !customFormat;
 
   const dateAsLocaleString = new Date(dateProp).toLocaleString(locale, {
     ...formatOptions,
     ...(hasTimeFormat && { timeStyle: timeFormat })
   });
+  const defaultDisplay = `${dateAsLocaleString}${displaySuffix ? ' ' + displaySuffix : ''}`;
```
Comment (mcoker) [inline packages/react-core/src/components/Timestamp/Timestamp.tsx:108]: Looks like maybe this what was causing the trailing whitespace? FWIW, it's still happening with custom content when you render `children`. 

I wonder if we could do something like what you had before but use something like [.trim()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/Trim) on the `display` var

```
const display = {!children ? `${dateAsLocaleString} ${timeZoneSuffix}` : children};
```

https://github.com/patternfly/patternfly-react/blob/8c299a46c78ddc5bea2aa325c47c29477f00baba/packages/react-core/src/components/Timestamp/Timestamp.tsx#L125

File: packages/react-core/src/components/Timestamp/Timestamp.tsx [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+export const Timestamp: React.FunctionComponent<TimestampProps> = ({
+  children,
+  className,
+  customFormat,
+  date: dateProp = new Date(),
+  dateFormat,
+  datetime,
+  displaySuffix = '',
+  is12Hour,
+  locale = undefined,
```
Comment (tlabaj) [inline packages/react-core/src/components/Timestamp/Timestamp.tsx:92]: you don't need the undefined here.

**Review by tlabaj — CHANGES_REQUESTED** [review]
Looking great Eric.  A few more comments.

File: packages/react-core/src/components/Timestamp/Timestamp.tsx [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+   */
+  children?: React.ReactNode;
+  /** Class to add to the outer span. */
+  className?: string;
+  /** Applies custom formatting to the displayed date and/or time. Passing this prop in will
+   * override the dateFormat and timeFormat props.
+   */
+  customFormat?: { [key: string]: string };
+  /** Determines the default content that is displayed inside the timestamp. */
+  date?: Date;
```
[the remark this answers is not in the document]
Comment (thatblindgeye) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:44]: I think we could. Originally I was under the impression that the datetime attribute should be a local date time string, but per MDN since a UTC timestring is valid I'd be good with removing the `datetime` prop and making that logic internal.

File: packages/react-core/src/components/Timestamp/__tests__/Timestamp.test.tsx [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+});
+
+test('Renders with 24 hour time when is12Hour is set to false for US locale', () => {
+  render(
+    <Timestamp
+      datetime={new Date(2022, 0, 1, 13, 0).toISOString()}
+      date={new Date(2022, 0, 1, 13, 0)}
+      is12Hour={false}
+    />
+  );
```
Comment (wise-king-sullyman) [inline packages/react-core/src/components/Timestamp/__tests__/Timestamp.test.tsx:121]: Should these tests be using the `locale` prop?

Comment (wise-king-sullyman) [reply packages/react-core/src/components/Timestamp/__tests__/Timestamp.test.tsx:121]: Also could you add a test for the `locale` prop itself?

**Review by wise-king-sullyman — CHANGES_REQUESTED** [review]
This looks really great 🚀 

I just have a few small testing things then I'll be ready to approve!

File: packages/react-core/src/components/Timestamp/Timestamp.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (tlabaj) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:44]: I think that would simplify the API.

File: packages/react-core/src/components/Timestamp/Timestamp.tsx (hunk 5 above)
Comment (thatblindgeye) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:108]: So using trim() works for the default display, but since the children prop type is ReactNode, it doesn't work as easily. Updating the children type to just a string would resolve that (and ideally only text would get passed in as custom content since that is rendered inside `time` element, with the possibility of icons in the future being rendered outside the `time` element). @tlabaj wdyt about having the children prop only allow strings?

Comment (mcoker) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:108]: Ah bummer. The [<time> element accepts phrasing content](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/time) so IMO, it would be good to let a user put whatever they want in there, I suppose. In that case, it would be up to them to trim the whitespace if they're putting content around it where the border might extend past the timestamp text if they have a tooltip. It's an edge case anyways. I'm curious what @tlabaj thinks, too.

File: packages/react-core/src/components/Timestamp/__tests__/Timestamp.test.tsx (hunk 8 above)
Comment (thatblindgeye) [reply packages/react-core/src/components/Timestamp/__tests__/Timestamp.test.tsx:121]: Ah good catch. Updated these tests and added some tests for the `locale` prop!

File: packages/react-core/src/components/Timestamp/Timestamp.tsx (hunk 5 above)
Comment (tlabaj) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:102]: We probably want to keep the type ReactNode for the reasons @mcoker mentions above. I definitely think we want the default to trim the whitespace.  Maybe it is ok for the consumer to have to format any custom display on their own... I just wonder if custom display may be a common use case.  We would probably at least want to have the example demonstrate trimming the white space.

Comment (thatblindgeye) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:102]: Michael and I had talked a bit on Slack about this, and another alternative is updating the styles in Core so that `pf-c-timestamp` has `display: inline-block` which resolved that trailing whitespace issue when I applied the style inline in my local React build.

Comment (tlabaj) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:102]: @mcoker @thatblindgeye  even better!

File: packages/react-core/src/components/Timestamp/Timestamp.tsx [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+   * any props that format or add to displayed content.
+   */
+  children?: React.ReactNode;
+  /** Class to add to the outer span. */
+  className?: string;
+  /** Applies custom formatting to the displayed date and/or time. Passing this prop in will
+   * override the dateFormat and timeFormat props.
+   */
+  customFormat?: { [key: string]: string };
+  /** Determines the default content that is displayed inside the timestamp. If a dateTime prop is
```
Comment (tlabaj) [inline packages/react-core/src/components/Timestamp/Timestamp.tsx:43]: It looks like you need to update the prop description here

File: packages/react-core/src/components/Timestamp/__tests__/Timestamp.test.tsx [hunk 10]
```diff
@@ -80,92 +71,85 @@ test('Renders with custom formatting when customFormat is passed in', () => {
...
+  render(<Timestamp date={new Date(2022, 1, 1, 13, 0)} locale="en-GB" is12Hour />);
+
+  expect(screen.getByText('01/02/2022, 1:00:00 pm')).toBeInTheDocument();
+});
+
+test('Renders with default class names', () => {
+  render(<Timestamp date={new Date('1 Jan 2022 00:00:00 EST')} />);
+
+  expect(screen.getByText('1/1/2022, 12:00:00 AM')).toHaveClass('pf-c-timestamp__text');
+  expect(screen.getByText('1/1/2022, 12:00:00 AM').parentElement).toHaveClass('pf-c-timestamp');
```
Comment (wise-king-sullyman) [inline packages/react-core/src/components/Timestamp/__tests__/Timestamp.test.tsx:124]: 🤔 I've been doing one class assertion per test so far, but this approach could help make testing easier without *much* drawback. The only issue I can think of with this approach is that (because there are multiple points of failure for the test) seeing it fail provides *slightly* less value as it would require further investigation to see which class isn't being applied.

I would be curious to hear others' thoughts about these two approaches to testing for default class names.

File: packages/react-core/src/components/Timestamp/Timestamp.tsx [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+  locale,
+  timeFormat,
+  tooltip,
+  ...props
+}: TimestampProps) => {
+  const hasTimeFormat = timeFormat && !customFormat;
+  const formatOptions = {
+    ...(dateFormat && !customFormat && { dateStyle: dateFormat }),
+    ...(customFormat && { ...customFormat }),
+    ...(is12Hour !== undefined && { hour12: is12Hour })
```
Comment (tlabaj) [inline packages/react-core/src/components/Timestamp/Timestamp.tsx:95]: maybe just default is12Hour to false?

**Comment by tlabaj** [comment]
> Looks awesome to me! 2 things:
> 
> * Looks like we're missing the `__text` class on the `time` element. I'm assuming because that isn't showing up in react-styles since the class doesn't have any styles from core? In those instances, I believe what we do is hardcode the class so that it's there and we can reference it if we need without introducing a new class, but maybe one of the react devs can verify if that's still the case.
I missed this comment earlier.  Yeas @mcoker is correct and it looks like you already made the update.

**Review by tlabaj — CHANGES_REQUESTED** [review]
Looking great!  A couple of small comments.

**Review by wise-king-sullyman — APPROVED** [review]
🚀 🚀 🚀 

Not a blocker, but I would appreciate thoughts on the item I brought up below:

File: packages/react-core/src/components/Timestamp/Timestamp.tsx (hunk 11 above)
Comment (thatblindgeye) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:95]: So originally I left it without a default value since in that case, the hour format could be determined by the locale, and then if consumers want to override that they can by passing is12Hour in. I don't think I really feel strongly either way, so if it'd be better to add a default value for the prop I'm all for it (maybe a default of `true` so that the hours get set to a 12 hour format?)

File: packages/react-core/src/components/Timestamp/Timestamp.tsx (hunk 9 above)
Comment (thatblindgeye) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:43]: Context for this description is I wanted to convey that consumers could pass in their own dateTime prop to the component to override what the component itself constructs internally. Re-reading it it can be a bit confusing since there's no dateTime prop declared, and that could possibly be something better suited for a11y or design docs. How does the following sound instead?

"Determines the default content that is displayed inside the timestamp and the value of the inner time element's datetime attribute"

File: packages/react-core/src/components/Timestamp/__tests__/Timestamp.test.tsx (hunk 10 above)
Comment (thatblindgeye) [reply packages/react-core/src/components/Timestamp/__tests__/Timestamp.test.tsx:124]: That's a fair point! At least locally when testing *only* the timestamp suite, it does provide that context of what class isn't being applied. Running all of the tests locally definitely doesn't provide that context, and whether it would when checks are run in the PR could go either way.

For sake of clarity I'd be good with separating this to two tests (with components that have a *lot* of classes it *might* cause a bit of bloat to test each class individually, but since it's only 2 classes here I don't think it's really an issue).

File: packages/react-core/src/components/Timestamp/Timestamp.tsx (hunk 9 above)
Comment (tlabaj) [reply packages/react-core/src/components/Timestamp/Timestamp.tsx:43]: I think that is better.

**Review by tlabaj — APPROVED** [review]
LGTM.  Nice work.  The only comment I have that is not a blocker. is that may the ordering of examples should switch. I would expect the tooltip ones to be one after the other.  I will defer to @mcarrano on that though.

**Review by mcoker — APPROVED** [review]
🎃👍

**Review by mcarrano — APPROVED** [review]
Looks great. Thanks @thatblindgeye !

**Review by mmenestr — APPROVED** [review]
Agree looks great!!

**Review by nicolethoen — APPROVED** [review]
Keyboard and VO navigation both look beautiful. No new a11y violations in the docs or examples :)

LGTM
